File: ruby/lib/runapi/suno/resources/generate_persona.rb
# frozen_string_literal: true

module RunApi
  module Suno
    module Resources
      class GeneratePersona
        include RunApi::Core::ResourceHelpers

        ENDPOINT = "/api/v1/suno/generate_persona"
        RESPONSE_CLASS = Types::GeneratePersonaResponse

        def initialize(http)
          @http = http
        end

        def run(**params)
          params = compact_params(params)
          validate_params!(params)
          request(:post, ENDPOINT, body: params)
        end

        private

        def validate_params!(params)
          Validators.validate_generate_persona!(params, self)
        end
      end
    end
  end
end

File: python/src/runapi/suno/resources/generate_persona.py
"""Suno generate_persona resource (synchronous)."""

from __future__ import annotations

from typing import Any, Dict

from runapi.core import Resource

from .. import _validators
from ..types import GeneratePersonaResponse


class GeneratePersona(Resource):
    """Generate a reusable persona. Synchronous: run() returns the result directly."""

    ENDPOINT = "/api/v1/suno/generate_persona"

    RESPONSE_CLASS = GeneratePersonaResponse

    def run(self, **params: Any) -> Any:
        """Create a reusable style or voice persona (synchronous).

        Args:
            **params: persona parameters.

        Returns:
            The result.
        """
        compacted = self._compact_params(params)
        self._validate_params(compacted)
        return self._request("post", self.ENDPOINT, body=compacted)

    def _validate_params(self, params: Dict[str, Any]) -> None:
        _validators.validate_generate_persona(params)
